decidim-core/app/presenters/decidim/admin_log/user_presenter.rb

```diff
@@ -23,6 +23,23 @@ def action_string
         end
       end
 
+      def resource_presenter
+        @resource_presenter ||=
+          if action == "block" && action_log.resource.respond_to?(:user_name)
```

Comment:
💬 Uf, this one was hard to follow 😅. I feel it a bit hacky, and I think I found a way to implement it simpler. Let me explain it and tell me if you think if it's an improvement. If not, I can merge it as it is. My idea is to change this method with:

```ruby
def resource_presenter
  @resource_presenter ||= Decidim::ActionLog::UserResourcePresenter.new(action_log.resource, h, action_log.extra["resource"])
end
```

and then implement the `Decidim::ActionLog::UserResourcePresenter` class like this:
```ruby
module Decidim
  module ActionLog
    class UserResourcePresenter > Decidim::Log::ResourcePresenter
      # Private: Presents user name, even if it's blocked.
      #
      # Returns an HTML-safe String.
      def present_resource_name
        h.translated_attribute resource.user_name.presence || extra["title"]
      end
    end
  end
end
```

This way, all the `ActionLog` entries will be updated if the user is blocked or changes it's name, but I don't know if that's a good thing or not. Anyway, it can be changed to adapt it to the exact requirements we define, and I think is clearer and more prepared for the future. What do you think?

Reply:
OK but I'm just not sure if `resource` is always an instance of `Decidim::User` when the above would fail.

Also, this would change the current behavior of the other actions than "block" as well.

But I think neither of those points doesn't matter really.

Reply:
@leio10 Those changes are now done but as said, the resource is not necessarily always a user.

Reply:
Mmm, I asume that the resource will always be a User, as the class is called `UserPresenter`.

Regarding the type of action, I think you are right, I'm not completely sure if the change can affect other actions. But AFAIK, only the blocked users have a value for `user_name`, so it shouldn't change it so much. As the `UserResourcePresenter` doesn't know anything about the action type it cannot decide when to use the `user_name` and when not to do it.

Based on this I can imagine three options here:
* Leave it as it is now. In this case, the only "problem" that I see is when you perform an action, change your user name and then be blocked. In that case, the first action will appear in the log with the changed username, instead of showing the original username that is in `extra["title"]`, as it should do it now. It could also be a bit slower when showing several log actions, as maybe it needs to fetch the resource from the database (I don't know if it's done now).
* Modify the `UserResourcePresenter` to use the `user_name` only when the `extra["title"]` is equal to "Blocked user". This is too hacky for me, and I can't imagine what will happen when your username is "Blocked user", but it seems to be fast and simple.
* Modify the `UserPresenter` to use the `UserResourcePresenter` only when the `action` is `"block"`, similar to the `if` you had before. After thinking a bit, I think this is the best option (or the "less bad").

Sorry for wasting a bit more of your time, please feel free to choose the one you like the most and this time I will merge it as you decide. 🙏 😬

Reply:
OK.

I'm also not sure if `resource` can be anything else than the log entry's associated user resource object.

In the current spec that is testing `UserPresenter`, an instance of `Decidim::Organization` is provided as the resource:
https://github.com/decidim/decidim/blob/15d222fd367aca7c8dac7c84ad3337a90112d0ce/decidim-core/spec/presenters/decidim/admin_log/user_presenter_spec.rb#L15

This just got me thinking it could be something else too.

Reply:
By the way, coauthorable resources can have `Decidim::Organization` as author when they are of official origin:
https://github.com/decidim/decidim/blob/15d222fd367aca7c8dac7c84ad3337a90112d0ce/decidim-core/lib/decidim/coauthorable.rb#L29

Looking at the core code, I don't find any case where it is provided as the resource for `Decidim::UserPresenter` but it is technically possible outside of core for models implemeting `Decidim::Coauthorable`.

Reply:
@leio10 The PR is now updated with a new approach as per your comment regarding the performance which I think is extremely important point.

What I ended up doing is that rather than customizing the presenters, I override the resource title during the action log creation. This also migrates old action log entries to store the original user names instead of "Blocked user".